### PR TITLE
cudnn: init 8.8.0; rewrite to assist maintainability

### PR DESCRIPTION
**Pull request by ConnorBaker**
###### Description of changes

Adds CUDNN 8.8.0.

Rewrites the `cudnn` package to move objects out of the nix files and into a more easily maintainable sorted JSON array.

Instead of hardcoding the default version of CUDNN for each CUDA release, it is now chosen based on the versions available in `releases.json`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ConnorBaker** [comment]
Result of `nixpkgs-review pr 217363` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
  <ul>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python311Packages.tensorrt</li>
  </ul>
</details>
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>cudaPackages.cudnn</li>
    <li>cudaPackages.cudnn_8_7_0</li>
    <li>cudaPackages.tensorrt</li>
    <li>katagoWithCuda</li>
    <li>python310Packages.TheanoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.tensorrt</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.TheanoWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
  </ul>
</details>

**Comment by ConnorBaker** [comment]
cc @NixOS/cuda-maintainers

**Review by samuela — COMMENTED** [review]
Thanks for contributing @ConnorBaker ! Overall I feel that this is a step forward. However, why use json instead of Nix for the releases info?

Also IIRC I originally wrote cudnn such that it had version constraints for each package like this, but we got rid of it at some point. I don't recall exactly why... I do remember there being a point where one of the CUDA releases required a cuDNN but there was no compatible version in the official table. Other distros were just winging it with whatever latest cudnn was and I think we ended up doing the same. I'm not opposed to this design, but might be worth checking the history to see why it was removed before.

**Comment by ConnorBaker** [comment]
Will also rewrite the releases as Nix here

File: pkgs/development/libraries/science/math/cudnn/releases.nix [hunk 1]
```diff
@@ -0,0 +1,194 @@
+[
+  {
+    version = "7.4.2.24";
+    minCudaVersion = "10.0";
+    maxCudaVersion = "10.0";
+    url = "https =//developer.download.nvidia.com/compute/redist/cudnn/v7.4.2/cudnn-10.0-linux-x64-v7.4.2.24.tgz";
+    hash = "sha256-Lt/IagK1DRfojEeJVaMy5qHoF05+U6NFi06lH68C2qM="
+    ;
```
Comment (samuela) [inline pkgs/development/libraries/science/math/cudnn/releases.nix:8]:
```suggestion
    hash = "sha256-Lt/IagK1DRfojEeJVaMy5qHoF05+U6NFi06lH68C2qM=";
```
applies to the other entries as well

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 217363` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages failed to build:</summary>
  <ul>
    <li>cudaPackages.cudnn (cudaPackages.cudnn_8_8_0)</li>
    <li>cudaPackages.tensorrt (cudaPackages.tensorrt_8_4_0)</li>
    <li>katagoWithCuda</li>
    <li>python310Packages.TheanoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python310Packages.tensorrt</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.TheanoWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.tensorrt</li>
  </ul>
</details>

**Comment by samuela** [comment]
I'm seeing the following error:
```
error: builder for '/nix/store/0sczii183mc2wkx9xlx5a7839cdymapj-cudnn-linux-x86_64-8.8.0.121_cuda11-archive.tar.xz.drv' failed with exit code 1;
       last 10 log lines:
       > Warning: Problem : timeout. Will retry in 1 seconds. 3 retries left.
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: https
       > Warning: Problem : timeout. Will retry in 2 seconds. 2 retries left.
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: https
       > Warning: Problem : timeout. Will retry in 4 seconds. 1 retries left.
       >   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: https
       >
       > trying =//developer.download.nvidia.com/compute/cudnn/redist/cudnn/linux-x86_64/cudnn-linux-x86_64-8.8.0.121_cuda11-archive.tar.xz
       > curl: (3) URL using bad/illegal format or missing URL
       > error: cannot download cudnn-linux-x86_64-8.8.0.121_cuda11-archive.tar.xz from any mirror
       For full logs, run 'nix log /nix/store/0sczii183mc2wkx9xlx5a7839cdymapj-cudnn-linux-x86_64-8.8.0.121_cuda11-archive.tar.xz.drv'.
```

**Comment by ConnorBaker** [comment]
Sorry, that was the result of a regex gone awry! Should be fixed now

**Comment by ConnorBaker** [comment]
Result of `nixpkgs-review pr 217363` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>cudaPackages.tensorrt (cudaPackages.tensorrt_8_4_0)</li>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python310Packages.tensorrt</li>
    <li>python311Packages.tensorrt</li>
  </ul>
</details>
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>cudaPackages.cudnn (cudaPackages.cudnn_8_8_0)</li>
    <li>katagoWithCuda</li>
    <li>python310Packages.TheanoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.TheanoWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
  </ul>
</details>

**Comment by ConnorBaker** [comment]
###### Failed derivations

- [ ] TensorRT-8.4.0.6.Linux.x86_64-gnu.cuda-11.6.cudnn8.3.tar.gz.drv (derivation hash: j5r8g41359g4jsw3cd2dg6ylkfcjyncm)
  - https://gist.github.com/686b281045957cc160cc2381981cb21c
- [ ] tensorflow-gpu-2.11.0.drv (derivation hash: cimd54qzsc0h4vyfr6jf465xsn4hc2hc)
  - https://gist.github.com/634c38995873d870eedc11d034f51fb3

**Comment by samuela** [comment]
Result of `nixpkgs-review pr 217363` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>cudaPackages.tensorrt (cudaPackages.tensorrt_8_4_0)</li>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python310Packages.tensorrt</li>
    <li>python311Packages.tensorrt</li>
  </ul>
</details>
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>cudaPackages.cudnn (cudaPackages.cudnn_8_8_0)</li>
    <li>katagoWithCuda</li>
    <li>python310Packages.TheanoWithCuda</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.TheanoWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
  </ul>
</details>

**Pull request merged**